File: MCCME/1497.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <set>
#include <map>
#include <string>
using namespace std;

int main() {
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);

    int a, b;
    cin >> a >> b;

    if (a) {
        if (b % a)
            cout << "no solution";
        else
            cout << b / a;
    } else {
        if (b)
            cout << "no solution";
        else
            cout << "many solutions";
    }
}

File: EOlymp/1497.py
n = int(input())

l, r = 1, 1
while r * r <= n:
    r *= 2

while l + 1 < r:
    m = l + (r - l) // 2
    if m * m <= n:
        l = m
    else:
        r = m

print(l)
